File: simple-cqrs/Command.cs
using System;
using LanguageExt;
using static LanguageExt.Prelude;

namespace SimpleCQRS
{
   public interface ICommand 
   {
      Guid CommandId { get; }
      Guid RootId { get; }
   }

   public abstract class ReadAggregate<T> : ICommand
   {
      public Guid CommandId { get; } = Guid.NewGuid();
      public Guid RootId { get; } = Guid.NewGuid();
      public AsyncValue<ReadCommandResult<T>> Result { get; } = new AsyncValue<ReadCommandResult<T>>();

      public ReadAggregate(Guid rootId) => 
         RootId = rootId;
   }

   public abstract class CreateRootCommand : ICommand
   {
      public Guid CommandId { get; } = Guid.NewGuid();
      public Guid RootId { get; } = Guid.NewGuid();
      public AsyncValue<CommandResult> Result { get; } = new AsyncValue<CommandResult>();
   }

   public abstract class Command : ICommand
   {
      public Guid CommandId { get; } = Guid.NewGuid();
      public Guid RootId { get; }
      public AsyncValue<CommandResult> Result { get; } = new AsyncValue<CommandResult>();
      public int OriginalVersion { get; }

      public Command(Guid rootId, int originalVersion)
      {
         RootId = rootId;
         OriginalVersion = originalVersion < 1
            ? throw new ArgumentException("originalVersion must be at least 1")
            : originalVersion;
      }
   }

   public abstract class DetailCommand : Command
   {
      public Guid DetailId { get; }

      public DetailCommand(Guid rootId, int originalVersion)
         : base(rootId, originalVersion)
      {
         DetailId = Guid.NewGuid();
      }

      public DetailCommand(Guid rootId, Guid detailId, int originalVersion)
         : base(rootId, originalVersion)
      {
         DetailId = detailId;
      }
   }
}

File: Fescq/Command.fs
module Fescq.Command

open Fescq
open System
open System.Reactive.Subjects
open System.Reactive.Linq;
open FSharpx.Control.Observable

type AsyncValue<'t> () = 
   let subject = new ReplaySubject<'t>()

   member x.Set(value:'t) =
      subject.OnNext(value)
      subject.OnCompleted()

   member x.Wait() =
      subject.LastAsync()
      |> Async.AwaitObservable
      |> Async.RunSynchronously


type AggregateWithHistory<'t> = {
   Aggregate: 't
   History: seq<Event>
}

type ICommand = 
   abstract member CommandId: Guid
   abstract member AggregateId: Guid

type CommandResult = {
   Id: Guid
   Result: Result<unit, string>
}

type ReadResult<'t> = {
   Id: Guid
   Result: Result<AggregateWithHistory<'t>, string>
}



[<AbstractClass>]
type CreateAggregateCommand () =
   member x.Result = new AsyncValue<CommandResult>()
   interface ICommand with
      member val CommandId = System.Guid.NewGuid() with get
      member val AggregateId = System.Guid.NewGuid() with get

[<AbstractClass>]
type ReadAggregateCommand<'t> (aggregateId:Guid) =
   member x.Result = new AsyncValue<ReadResult<'t>>()
   interface ICommand with
      member val CommandId = System.Guid.NewGuid() with get
      member val AggregateId = aggregateId with get

[<AbstractClass>]
type UpdateCommand (aggregateId:Guid, originalVersion:int) =

   let ver = 
      if originalVersion < 1 then 
         failwith "originalVersion must be at least 1"
      else
         originalVersion

   member x.Result = new AsyncValue<CommandResult>()
   member val OriginalVersion = ver with get
   interface ICommand with
      member val CommandId = System.Guid.NewGuid() with get
      member val AggregateId = aggregateId with get

[<AbstractClass>]
type DetailCommand (aggregateId:Guid, detailId:Guid, originalVersion:int) =  
   inherit UpdateCommand(aggregateId, originalVersion)
   member x.DetailId = detailId
   new (aggregateId:Guid, originalVersion:int) = DetailCommand(aggregateId, Guid.NewGuid(), originalVersion) 
